// @trpc/server/dist/unstable-core-do-not-import/stream/sse.mjs
import { getTRPCErrorFromUnknown } from '../error/TRPCError.mjs';
import { run, identity } from '../utils.mjs';
import { isTrackedEnvelope } from './tracked.mjs';
import { withCancel, takeWithGrace } from './utils/asyncIterable.mjs';
import { createDeferred } from './utils/createDeferred.mjs';
import { createReadableStream } from './utils/createReadableStream.mjs';
import { createPromiseTimer } from './utils/promiseTimer.mjs';
import { withPing, PING_SYM } from './utils/withPing.mjs';

const SERIALIZED_ERROR_EVENT = 'serialized-error';
/**
 *
 * @see https://html.spec.whatwg.org/multipage/server-sent-events.html
 */ function sseStreamProducer(opts) {
    const stream = createReadableStream();
    stream.controller.enqueue({
        comment: 'connected'
    });
    const { serialize =identity  } = opts;
    const ping = {
        enabled: opts.ping?.enabled ?? false,
        intervalMs: opts.ping?.intervalMs ?? 1000
    };
    run(async ()=>{
        let iterable = opts.data;
        iterable = withCancel(iterable, stream.cancelledPromise);
        if (opts.emitAndEndImmediately) {
            iterable = takeWithGrace(iterable, {
                count: 1,
                gracePeriodMs: 1
            });
        }
        let maxDurationTimer = null;
        if (opts.maxDurationMs != null && opts.maxDurationMs > 0 && opts.maxDurationMs !== Infinity) {
            maxDurationTimer = createPromiseTimer(opts.maxDurationMs).start();
            iterable = withCancel(iterable, maxDurationTimer.promise);
        }
        if (ping.enabled && ping.intervalMs !== Infinity && ping.intervalMs > 0) {
            iterable = withPing(iterable, ping.intervalMs);
        }
        try {
            for await (const value of iterable){
                if (value === PING_SYM) {
                    stream.controller.enqueue({
                        comment: 'ping'
                    });
                    continue;
                }
                const chunk = isTrackedEnvelope(value) ? {
                    id: value[0],
                    data: value[1]
                } : {
                    data: value
                };
                if ('data' in chunk) {
                    chunk.data = JSON.stringify(serialize(chunk.data));
                }
                stream.controller.enqueue(chunk);
            }
        } catch (err) {
            // ignore abort errors, send any other errors
            if (!(err instanceof Error) || err.name !== 'AbortError') {
                // `err` must be caused by `opts.data`, `JSON.stringify` or `serialize`.
                // So, a user error in any case.
                const error = getTRPCErrorFromUnknown(err);
                const data = opts.formatError?.({
                    error
                }) ?? null;
                stream.controller.enqueue({
                    event: SERIALIZED_ERROR_EVENT,
                    data: JSON.stringify(serialize(data))
                });
            }
        } finally{
            maxDurationTimer?.clear();
            stream.controller.close();
        }
    }).catch((err)=>{
        // should not be reached; just in case...
        stream.controller.error(err);
    });
    return stream.readable.pipeThrough(new TransformStream({
        transform (chunk, controller) {
            if ('event' in chunk) {
                controller.enqueue(`event: ${chunk.event}\n`);
            }
            if ('data' in chunk) {
                controller.enqueue(`data: ${chunk.data}\n`);
            }
            if ('id' in chunk) {
                controller.enqueue(`id: ${chunk.id}\n`);
            }
            if ('comment' in chunk) {
                controller.enqueue(`: ${chunk.comment}\n`);
            }
            controller.enqueue('\n\n');
        }
    }));
}
/**
 * @see https://html.spec.whatwg.org/multipage/server-sent-events.html
 */ function sseStreamConsumer(opts) {
    const { deserialize =(v)=>v , shouldRecreateOnError  } = opts;
    const signal = opts.signal;
    let eventSource = null;
    let lock = null;
    const stream = createReadableStream();
    function createEventSource(...args) {
        const es = new EventSource(...args);
        if (signal.aborted) {
            es.close();
        } else {
            signal.addEventListener('abort', ()=>es.close());
        }
        /**
     * Dispatch an event to the stream controller
     *
     * Will be a no-op if the event source has been replaced
     */ const dispatch = (fn)=>{
            run(async ()=>{
                while(lock){
                    await lock;
                }
                if (es === eventSource) {
                    await fn(stream.controller);
                }
            }).catch((error)=>{
                stream.controller.error(error);
            });
        };
        const pauseDispatch = async (fn)=>{
            while(lock){
                await lock;
            }
            if (es !== eventSource) {
                return;
            }
            const deferred = createDeferred();
            lock = deferred.promise;
            try {
                await fn();
            } finally{
                lock = null;
                deferred.resolve();
            }
        };
        es.addEventListener('open', ()=>{
            dispatch((controller)=>{
                controller.enqueue({
                    type: 'opened',
                    eventSource: es
                });
            });
        });
        es.addEventListener(SERIALIZED_ERROR_EVENT, (msg)=>{
            dispatch(async ()=>{
                if (shouldRecreateOnError) {
                    await pauseDispatch(async ()=>{
                        const recreate = await shouldRecreateOnError({
                            type: SERIALIZED_ERROR_EVENT,
                            error: deserialize(JSON.parse(msg.data))
                        });
                        if (recreate) {
                            await recreateEventSource();
                        }
                    });
                }
                dispatch((controller)=>{
                    controller.enqueue({
                        type: 'error',
                        error: deserialize(JSON.parse(msg.data)),
                        eventSource: es
                    });
                });
            });
        });
        es.addEventListener('error', (event)=>{
            dispatch(async ()=>{
                if (shouldRecreateOnError) {
                    await pauseDispatch(async ()=>{
                        const recreate = await shouldRecreateOnError({
                            type: 'event',
                            event
                        });
                        if (recreate) {
                            await recreateEventSource();
                        }
                    });
                }
                dispatch((controller)=>{
                    if (es.readyState === EventSource.CLOSED) {
                        controller.error(event);
                    } else {
                        controller.enqueue({
                            type: 'connecting',
                            eventSource: es
                        });
                    }
                });
            });
        });
        es.addEventListener('message', (msg)=>{
            dispatch((controller)=>{
                const chunk = deserialize(JSON.parse(msg.data));
                const def = {
                    data: chunk
                };
                if (msg.lastEventId) {
                    def.id = msg.lastEventId;
                }
                controller.enqueue({
                    type: 'data',
                    data: def,
                    eventSource: es
                });
            });
        });
        return es;
    }
    async function recreateEventSource() {
        eventSource?.close();
        const [url, init] = await Promise.all([
            opts.url(),
            opts.init()
        ]);
        eventSource = createEventSource(url, init);
        stream.controller.enqueue({
            type: 'connecting',
            eventSource
        });
    }
    recreateEventSource().catch(()=>{
    // prevent unhandled promise rejection
    });
    return {
        [Symbol.asyncIterator] () {
            const reader = stream.readable.getReader();
            const iterator = {
                async next () {
                    const value = await reader.read();
                    if (value.done) {
                        return {
                            value: undefined,
                            done: true
                        };
                    }
                    return {
                        value: value.value,
                        done: false
                    };
                },
                async return () {
                    reader.releaseLock();
                    return {
                        value: undefined,
                        done: true
                    };
                }
            };
            return iterator;
        }
    };
}
const sseHeaders = {
    'Content-Type': 'text/event-stream',
    'Cache-Control': 'no-cache, no-transform',
    'X-Accel-Buffering': 'no',
    Connection: 'keep-alive'
};

export { sseHeaders, sseStreamConsumer, sseStreamProducer };
